src/main/java/org/assertj/core/condition/VerboseCondition.java

```diff
@@ -0,0 +1,158 @@
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * Copyright 2012-2021 the original author or authors.
+ */
+package org.assertj.core.condition;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.function.BiPredicate;
+import java.util.function.Function;
+
+import org.assertj.core.annotations.Beta;
+import org.assertj.core.api.Condition;
+
+/**
+ * Condition that shows the expected and the value to test in the verbose description.
+ * Part of the resulting description is also a verbal description of BiPredicate. 
+ * The description output of the expected and the value to test could also be transformed by functions.
+ *
+ * <pre><code class='java'> Condition&lt;String&gt; verboseCondition = VerboseCondition.verbose(
+        4, // the expected value
+        (String actual, Integer expected) -&gt; actual.length() &lt; expected, // matchesBiPredicate
+        "word must be shorter than", // verbal description of the matchesBiPredicate
+        (i) -&gt; i + " (maximum word length)", // expected value transformation
+        (s) -&gt; String.format("%s (original word: %s)", s.length(), s)); // actual value transformation
+ *</code></pre>
+ * Tested against the (not matching) word `foooo` the description of the condition would be 
+ * `word must be shorter than &lt;4 (maximum word length)&gt; but was &lt;5 (original word: foooo)&gt;`
+ *<p>
+ * @param <T> the type of object the given condition accept.
+ * @param <EXPECTED> the type of object the expected value has.
+ *
+ * @author Stefan Bischof
+ */
+@Beta
+public final class VerboseCondition<T, EXPECTED> extends Condition<T> {
+
+  private EXPECTED expectedValue;
+
+  private BiPredicate<T, EXPECTED> matchBiPredicate;
+
+  private String matchDescription;
+
+  private Function<T, ?> actualValueTransformation;
+
+  private Function<EXPECTED, ?> expectedValueTransformation;
+
+
+  /**
+   * Creates a new <code>{@link VerboseCondition}</code>
+   *
+   * @param <T> the type of object the given condition accept.
+   * @param <EXPECTED> the type of object the expected value has.
+   * @param expectedValue the object the given value would be tested against.
+   * @param matchesBiPredicate the BiPredicate that tests the value to test against the expected value.
+   * @return the created {@code VerboseCondition}.
+   * @throws NullPointerException if the matchesBiPredicate is {@code null}.
+   */
+  public static <T,EXPECTED> VerboseCondition<T,EXPECTED> verbose(EXPECTED expectedValue, BiPredicate<T, EXPECTED> matchesBiPredicate) {
```

Comment:
I would rename it to `verboseCondition` so that a static import would read nicely

Reply:
renamed